src/app/qgshandlebadlayers.cpp

```diff
@@ -364,79 +370,48 @@ void QgsHandleBadLayers::editAuthCfg()
 void QgsHandleBadLayers::apply()
 {
   QgsProject::instance()->layerTreeRegistryBridge()->setEnabled( true );
-  QHash<QString, QString> baseChange;
-
 
   for ( int i = 0; i < mLayerList->rowCount(); i++ )
   {
     int idx = mLayerList->item( i, 0 )->data( Qt::UserRole ).toInt();
     QDomNode &node = const_cast<QDomNode &>( mLayers[ idx ] );
 
-    const QString name = mLayerList->item( i, 0 )->text();
     QTableWidgetItem *item = mLayerList->item( i, 4 );
     QString datasource = item->text();
+    QString fileName;
+    const QString name { mLayerList->item( i, 0 )->text() };
```

Comment:
Would it be possible to use the layer id instead of the name and save it as data (something like `mLayerList->item( i, 0 )->data(QgsHandleBasdLayers::LayerIdRole).toString()`)?

Reply:
Possibly, would this be robust enough to resist changes to the layer list? The idea is mostly to have the data associated to the layer and not have changes to the table (removing rows, etc) affect anything. If you tell me that the layerid will be maintained I don't have an issue with this.

Reply:
In short yes.
Slightly longer: The `item` is created once in the code and then managed and moved around by the table (e.g. sorting) until the row is removed again. So any `item->setData` is safe and persistent. I'd always attach data that relates tot he whole row to the first column (`item( i , 0 )`) to make it easy to obtain again.

Reply:
Would `node.namedItem( QStringLiteral( "id" ) ).toElement().text()` be fine, as this is already in use.

Reply:
Yes, that would be fine :+1:

Reply:
Keep in mind that `mapLayer->setDataSource( datasource, name, provider, options );` relies on the name and using the id forces me to use one more variable.

Reply:
Ping @elpaso , I think you are most familiar with this code

Reply:
You shouldn't need to change setDataSource here -- that doesn't require an ID, as the ID should never change (even after changing the layer's data source, it must still be left untouched)

Reply:
I don't plan to alter setDataSource, I simply wanted to say that relying on the name instead of the id is not removing the need to create the name variable.

Reply:
I have implemented something, using UserRole + 2 (as +0 is already used)

Reply:
Ah, I misinterpreted, sorry